mspsds-web/deploy.sh

```diff
@@ -8,27 +8,54 @@ set -ex
 # The caller should have the following environment variables set:
 #
 # SPACE: the space to which you want to deploy
-# If NO_START is set the app won't be started
 
 DOMAIN=london.cloudapps.digital
 HOSTNAME=mspsds-$SPACE
-COMPONENT=mspsds-web
+APP=mspsds-web
+APP_PREEXISTS=$(cf app $APP && echo 0 || echo 1)
 
-cf app $COMPONENT || APP_DOES_NOT_EXIST="true"
+if [[ $APP_PREEXISTS ]]; then
+    # We should deploy to a temporary location such that we can do a blue-green deployment
+    $NEW_HOSTNAME=$HOSTNAME-temp
+    $NEW_APP=$APP-temp
+else
+    # We don't need to deploy to a temporary location
+    echo "No existing app found. Performing first-time setup."
+    $NEW_HOSTNAME=$HOSTNAME
+    $NEW_APP=$APP
+fi
 
-rm -rf ./$COMPONENT/vendor/shared-web/
-cp -a ./shared-web/. ./$COMPONENT/vendor/shared-web/
+# Copy the environment helper script
+cp -a ./infrastructure/env/. ./mspsds-web/env/
 
-if [[ ! $APP_DOES_NOT_EXIST ]]; then
-    # Route to the maintenance page
-    cf map-route maintenance $DOMAIN --hostname $HOSTNAME
-    cf unmap-route $COMPONENT $DOMAIN --hostname $HOSTNAME
-fi
+# Copy in the shared dependencies
+rm -rf ./$APP/vendor/shared-web/
+cp -a ./shared-web/. ./$APP/vendor/shared-web/
 
-cf push -f ./$COMPONENT/manifest.yml $( [[ $APP_DOES_NOT_EXIST ]] && printf %s "--hostname $HOSTNAME" || printf %s "--no-route" ) $( [[ $NO_START ]] && printf %s "--no-start" )
+# Deploy the new app, set the hostname and start the app
+cf push $NEW_APP -f ./$APP/manifest.yml --hostname $NEW_HOSTNAME --no-start
+cf set-env $NEW_APP MSPSDS_HOST "https://$NEW_HOSTNAME.$DOMAIN"
+cf start $NEW_APP
 
-if [[ ! $APP_DOES_NOT_EXIST ]]; then
-    # Route to newly deployed app
-    cf map-route $COMPONENT $DOMAIN --hostname $HOSTNAME
-    cf unmap-route maintenance $DOMAIN --hostname $HOSTNAME
+
+if [[ ! $APP_PREEXISTS ]]; then
+    # We don't need to go any further
+    exit 0
 fi
+
+# TODO smoke test or manual confirmation?
+
+# Unmap the temporary hostname from the new app
+cf unmap-route $NEW_APP $DOMAIN --hostname $NEW_HOSTNAME
+
+# Map the live hostname to the new app
+cf set-env $NEW_APP MSPSDS_HOST "https://$HOSTNAME.$DOMAIN"
+cf restage $NEW_APP
```

Comment:
If the environment variable isn't used at build time, it's safe to `restart` rather than `restage`
```suggestion
cf restart $NEW_APP
```